### PR TITLE
fixes #85, i. e. error code -166...

### LINKED ISSUE #85: error code -1160725808
Occasionally, the following bug appears. No idea so far what is the reason for it.

```
File "pyndl/ndl_parallel.pyx", line 65, in pyndl.ndl_parallel.learn_inplace (pyndl/ndl_parallel.c:2085)
OSError: binary files does not have proper format, error code -1160725808
```

### PR DESCRIPTION
**Pull request by derNarr**
Fixes #85 .

Changes proposed in this pull request:
- if the number of all oucomes was a multiple of the chucksize, which defaults to 10, multithreading tried to compute an additional job package without any outcomes in it.
- This is fixed by checking if there is a remainder and only computing the last job package when there is a non-zero remainder.

As the ndl function call returned with an error, all calculated results so far are unaffected.

### CODE REVIEW

File: pyndl/ndl_parallel.pyx [hunk 1]
```diff
@@ -52,29 +52,39 @@ def learn_inplace(binary_file_paths, np.ndarray[dtype_t, ndim=2] weights,
     cdef unsigned int length_all_outcomes = all_outcomes.shape[0]
     cdef char* fname
     cdef unsigned int start_val, end_val, ii, number_parts
-    cdef int error = 4
+    cdef int error = 3
+    # error codes:
+    #  0: no error
+    #  1: magic number does not match
+    #  2: version number does not match
+    #  3: error is never written
```
Comment (dekuenstle) [inline pyndl/ndl_parallel.pyx]: In 'never written' the same as 'undefined', meaning: There was an error, but we don't have any details.

File: pyndl/ndl_parallel.pyx [hunk 2]
```diff
@@ -52,29 +52,39 @@ def learn_inplace(binary_file_paths, np.ndarray[dtype_t, ndim=2] weights,
...
-            break
-          error = 0
-          error = learn_inplace_ptr(fname, weights_ptr, mm, alpha, beta1,
-                            beta2, lambda_, all_outcomes_ptr, start_val,
-                            end_val)
+        filename_byte_string = binary_file_path.encode("UTF-8")
+        fname = filename_byte_string
+
+        number_parts = (length_all_outcomes // chunksize)
+        if length_all_outcomes % chunksize != 0:
```
Comment (dekuenstle) [inline pyndl/ndl_parallel.pyx]: could also be a oneliner:   <int>ceil(length_all_outcomes / chunksize)

File: pyndl/ndl_parallel.pyx [hunk 3]
```diff
@@ -52,29 +52,39 @@ def learn_inplace(binary_file_paths, np.ndarray[dtype_t, ndim=2] weights,
     cdef unsigned int length_all_outcomes = all_outcomes.shape[0]
     cdef char* fname
     cdef unsigned int start_val, end_val, ii, number_parts
-    cdef int error = 4
+    cdef int error = 3
+    # error codes:
```
Comment (dekuenstle) [inline pyndl/ndl_parallel.pyx]: would be great, if we could use constants / enum for the error codes, this  would improve readability

File: pyndl/ndl_parallel.pyx [hunk 4]
```diff
@@ -88,31 +98,33 @@ def learn_inplace_2(binary_file_paths, np.ndarray[dtype_t, ndim=2] weights,
     cdef unsigned int length_all_outcomes = all_outcomes.shape[0]
     cdef char* fname
     cdef unsigned int start_val, end_val
-    cdef int error = 4
+    cdef int error = 3
```
Comment (dekuenstle) [inline pyndl/ndl_parallel.pyx]: Better use constant / enum

File: pyndl/ndl_parallel.pyx (hunk 1 above)
Comment (derNarr) [reply pyndl/ndl_parallel.pyx]: It is the initialized value for error. Therefore no more information is available except, for that the variable was initialized with 3 and then returned.

Comment (derNarr) [reply pyndl/ndl_parallel.pyx]: Do you have any idea how to phrase this better? Is there a standard way of doing this? I did not want to initialize the error variable with 0, therefore I picked a random non-zero number.

**Comment by derNarr** [comment]
I integrated the `math.ceil` code, but my enum integration did not worked. This is a very good suggestion that improves readability. I will make a separate pull request for the enum.

File: pyndl/ndl_parallel.pyx (hunk 1 above)
Comment (dekuenstle) [reply pyndl/ndl_parallel.pyx]: Should be fine the way it is here. Readability / naming will improve by PR #173

**Pull request merged**